### PR TITLE
vmware_guest: Rename existing VM, or mark as template

### PR DESCRIPTION
**Pull request by aperigault**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION
```
2.2.0
```

##### SUMMARY
- Add capability to mark an existing VM as template
- Add capability to rename an existing VM

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -1399,6 +1404,21 @@ def reconfigure_vm(self):
 
             change_applied = True
 
+        # Mark VM as Template
+        if self.params['is_template']:
+            task = self.current_vm_obj.MarkAsTemplate()
+
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: can you remove this empty line ?

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Comment by nerzhul** [comment]
Except the empty line it's good for me

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -51,6 +51,11 @@
         description:
             - Name of the newly deployed guest
         required: True
+   new_name:
+        description:
+            - Rename a VM
+            - New Name of the exising VM
+        required: False
```
Comment (gundalow) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:58]: Please add the following line, at the same indentation level as `required: False`
`version_added: "2.3`

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 2 above)
Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py:58]: Fixed!

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 1 above)
Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Fixed!

**Review by nerzhul — APPROVED** [review]
Okay now, thanks

**Comment by nerzhul** [comment]
shipit

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged, thanks

**Comment by dagwieers** [comment]
Hmm, that's not a good interface for doing an operation IMO.
What happens if you run this task twice ?
What about other operations, like deleting a VM, moving it to another location, ...
IMO all these actions do not belong in vmware_guest.

**Comment by aperigault** [comment]
@dagwieers What is your proposal ?
If I extend checks it will ok for you ?

**Comment by nerzhul** [comment]
@dagwieers if you run twice you will not found original VM then changed = False, it's lgtm

**Comment by nerzhul** [comment]
@aperigault can you add a bugfix to check if vm exists when new_name is provided ? If not you should fail

**Comment by aperigault** [comment]
@nerzhul Yes I'm writing this check right now. I also add an idempotent check if template is already one.

**Comment by nerzhul** [comment]
good point

**Comment by dagwieers** [comment]
Again, I think it is a bad design to mix states like absent/present, with operations like moving just because there is another parameter new_name. If you add more parameters for other operations (just like the snapshotting) it becomes a big problem IMO.

And adding more checks internally is hell. Because you will have to explain the user the different scenarios and what can work together and what not.

**Comment by dagwieers** [comment]
That's the reason why I would like to move facts gathering and snapshot operations out of vmware_guest. Not adding more things.

**Comment by aperigault** [comment]
Your proposal is to do a new module only to rename a VM ? And an other one to Mark Vm as template ?

**Comment by nerzhul** [comment]
@aperigault doesn't make sense to vm name and vm is_template as they are attribute of the VM, but it make sense for snapshot (which is an object itself) and for facts gathering (like ec2_facts)

**Comment by aperigault** [comment]
@nerzhul I'm agree. I'm waiting for @dagwieers proposal.

**Comment by dagwieers** [comment]
No, what I am saying is that the existence of parameters is not a good way to indicate operations.
That will become a big mess quickly. We have `state:` for operations.

That said, I do think that some operations do not belong in vmware_guest. I would like to see how other modules do the same thing (and even if they use the existence of `new_name:`  as an operation, that's bad.

**Comment by aperigault** [comment]
@dagwieers Is not exactly the same thing but there is another rename function here : https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/lxd/lxd_profile.py#L153 implemented exactly in the same way. This does not mean that it is good...
For now I do a new commit to add checks : https://github.com/aperigault/ansible/commit/d600d5af8f297cdb41bdf0eff66dc88a50fb8da3 I'm waiting for your opinion before do a new PR.
@bcoca Have you an opinion about this ?

**Comment by dagwieers** [comment]
BTW the reason why I am pedantic about this is because when we create something new the interface is more important than the code. The code we can change and fix afterwards, the interface not so easily.

That's why I am always very concerned when new things are added to the interface (we will have to support it from release onwards despite improved insights/understanding)

**Comment by bcoca** [comment]
@dagwieers  +10

**Comment by bcoca** [comment]
since name is not unique identifier i would overload name: and if it is different for the uiid: provided (actual unique identifier), change it.

If you don't know the uuid a previous task that looks it up should be added.

**Comment by dagwieers** [comment]
Ok, I like that. So something like this:

```yaml
- name: Move the VM
  vmware_guest:
    uuid: 1234-....-abcd
    name: new-name
    state: rename
```
That would be idempotent on consecutive runs, and uses `state:` for operations.

**Comment by aperigault** [comment]
I understand the problem but nobody use the uuid of a VM. If you think it's difficult to explain how to use the module with actual implementation. I think it would be more complicated to explain why you need to do this to rename a VM.
Instead of fail in renaming task your job will fail in previous task.
Can we use new_name only when state is rename for example ?

**Comment by bcoca** [comment]
i would even do it on 'state: present' as having a name not matching uuid would imply that imo, 'state=rename' just seems wrong.

@aperigault I think that having it as a 2 step process is fine, if you really dont like this approach i would suggest a dedicated module.

**Comment by dagwieers** [comment]
bcoca: I considered a rename a special operation that you wouldn't put in your normal (reran) playbooks. (Hence the idea of having it in special purpose module)

But what you suggest is even better if you consider a change in the playbook would automatically cause a rename. (the name would be just another property, which it actually is) It does mean that people initially make the right mapping (e.g. when moving van name-based to uuid-based handling in order to work with renames). Or they use a 2-step approach as you suggested.

It also makes uuid-based handling a preferred way of working, which is not bad either. I can live with this and feels much more in line with what @jctanner wanted too.

**Pull request merged**